mobilecoind-json/src/data_types.rs

```diff
@@ -261,6 +340,523 @@ impl From<&mc_mobilecoind_api::SendPaymentResponse> for JsonTransferResponse {
+        );
+
+        Ok(tx)
+    }
+}
+
+#[derive(Deserialize, Serialize)]
+pub struct JsonTxProposal {
+    pub input_list: Vec<JsonUnspentTxOut>,
+    pub outlay_list: Vec<JsonOutlay>,
+    pub tx: JsonTx,
+    pub fee: u64,
+    pub outlay_index_to_tx_out_index: Vec<(u64, u64)>,
+    pub outlay_confirmation_numbers: Vec<Vec<u8>>,
+}
+
+impl From<&mc_mobilecoind_api::TxProposal> for JsonTxProposal {
+    fn from(src: &mc_mobilecoind_api::TxProposal) -> Self {
+        let outlay_map: Vec<(u64, u64)> = src
+            .get_outlay_index_to_tx_out_index()
+            .iter()
+            .map(|(key, val)| (*key, *val))
+            .collect();
+        Self {
+            input_list: src
+                .get_input_list()
+                .iter()
+                .map(JsonUnspentTxOut::from)
+                .collect(),
+            outlay_list: src.get_outlay_list().iter().map(JsonOutlay::from).collect(),
+            tx: src.get_tx().into(),
+            fee: src.get_fee(),
+            outlay_index_to_tx_out_index: outlay_map,
+            outlay_confirmation_numbers: src.get_outlay_confirmation_numbers().to_vec(),
+        }
+    }
+}
+
+// Helper conversion between json and protobuf
+impl TryFrom<&JsonTxProposal> for mc_mobilecoind_api::TxProposal {
+    type Error = String;
+
+    fn try_from(src: &JsonTxProposal) -> Result<mc_mobilecoind_api::TxProposal, String> {
+        let mut inputs: Vec<mc_mobilecoind_api::UnspentTxOut> = Vec::new();
+        for input in src.input_list.iter() {
+            let utxo = mc_mobilecoind_api::UnspentTxOut::try_from(input)
+                .map_err(|err| format!("Failed to convert input: {}", err))?;
+            inputs.push(utxo);
+        }
+
+        let mut outlays: Vec<mc_mobilecoind_api::Outlay> = Vec::new();
+        for outlay in src.outlay_list.iter() {
+            let out = mc_mobilecoind_api::Outlay::try_from(outlay)
+                .map_err(|err| format!("Failed to convert outlay: {}", err))?;
+            outlays.push(out);
+        }
+
+        // Reconstruct the public address as a protobuf
+        let mut proposal = mc_mobilecoind_api::TxProposal::new();
+        proposal.set_input_list(RepeatedField::from_vec(inputs));
+        proposal.set_outlay_list(RepeatedField::from_vec(outlays));
+        proposal
+            .set_tx(Tx::try_from(&src.tx).map_err(|err| format!("Could not convert tx: {}", err))?);
+        proposal.set_fee(src.fee);
+        proposal.set_outlay_index_to_tx_out_index(HashMap::from_iter(
+            src.outlay_index_to_tx_out_index.clone(),
+        ));
+        proposal.set_outlay_confirmation_numbers(RepeatedField::from_vec(
+            src.outlay_confirmation_numbers.clone(),
+        ));
+
+        Ok(proposal)
+    }
+}
+
+#[derive(Deserialize, Serialize)]
+pub struct JsonGenerateTxRequest {
+    pub input_list: Vec<JsonUnspentTxOut>,
+    pub transfer: JsonReadRequestResponse,
```

Comment:
This is pretty hacky. The API should match the GRPC API, or it should be a different API.